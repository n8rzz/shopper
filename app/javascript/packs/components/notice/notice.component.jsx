import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

function buildClassNames(props) {
    return classnames({
        notice: true,
        'u-isVisuallyHidden': props.message == null || props.isVisible === false,
    });
}

export function Notice(props) {
    return (
        <div className={buildClassNames(props)}>
            {props.message}
        </div>
    );
}

Notice.propTypes = {
    message: PropTypes.string,
    isVisible: PropTypes.bool.isRequired,
};
