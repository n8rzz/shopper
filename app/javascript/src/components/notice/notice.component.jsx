import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export function Notice(props) {
    const noticeClassnames = classnames({
        notice: true,
        'mix-notice_isAlert': props.isAlert,
        'u-isVisuallyHidden': props.message === '' || !props.isVisible,
    });

    return (
        <div className={noticeClassnames}>
            {props.message}
        </div>
    );
}

Notice.propTypes = {
    message: PropTypes.string.isRequired,
    isAlert: PropTypes.bool.isRequired,
    isVisible: PropTypes.bool.isRequired,
};
