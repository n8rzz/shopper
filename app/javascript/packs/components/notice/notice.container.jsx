import React from 'react';
import PropTypes from 'prop-types';
import { Notice } from './notice.component';

const DEAFULT_REMOVAL_DELAY = 7000;

export class NoticeContainer extends React.Component {
    constructor(props) {
        super(props);

        this._timer = -1;
        this._updateNoticeVisibilityHandler = this._updateNoticeVisibility.bind(this);
        this.state = {
            isVisible: props.message !== null,
        };
    }

    componentDidMount() {
        if (this.props.message == null) {
            return;
        }

        this._registerRemovalTimer();
    }

    componentWillUnmount() {
        clearTimeout(this._timer);
    }

    _registerRemovalTimer() {
        this._timer = setTimeout(this._updateNoticeVisibilityHandler, DEAFULT_REMOVAL_DELAY);
    }

    _updateNoticeVisibility() {
        this.setState((prevState) => ({ isVisible: !prevState.isVisible }));
    }

    render() {
        return (
            <div
                className={'js-notice'}
            >
                <Notice message={this.props.message} isVisible={this.state.isVisible} />
            </div>
        );
    }
}

NoticeContainer.propTypes = {
    message: PropTypes.string,
};
