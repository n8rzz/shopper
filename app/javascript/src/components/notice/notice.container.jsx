import React from 'react';
import PropTypes from 'prop-types';
import Sticky from 'react-stickynode';
import classnames from 'classnames';
import EventService from '../../service/event.service';
import { Notice } from './notice.component';
import { EVENT_NAME } from '../../constants/event-names';

const DEAFULT_REMOVAL_DELAY = 7000;

export class NoticeContainer extends React.Component {
    constructor(props) {
        super(props);

        this._timer = -1;
        this._updateNoticeVisibilityHandler = this._updateNoticeVisibility.bind(this);
        this._onTriggerNoticeHandler = this._onTriggerNotice.bind(this);
        this._onStickyStateChangeHandler = this._onStickyStateChange.bind(this);
        this.state = {
            message: props.message,
            stickyStatus: 0,
        };

        EventService.on(EVENT_NAME.NOTICE_SUCCESS, this._onTriggerNoticeHandler);

        if (this.props.message != null) {
            this._registerRemovalTimer();
        }
    }

    componentWillUnmount() {
        clearTimeout(this._timer);
        EventService.off(EVENT_NAME.NOTICE_SUCCESS, this._onTriggerNoticeHandler);
    }

    _buildClassnames() {
        return classnames({
            'js-notice': true,
            'mix-stickyHeader_isSticky': this.state.message !== '' && this.state.stickyStatus === 2,
        });
    }

    _registerRemovalTimer() {
        if (this._timer !== -1) {
            clearTimeout(this._timer);
        }

        this._timer = setTimeout(this._updateNoticeVisibilityHandler, DEAFULT_REMOVAL_DELAY);
    }

    _onStickyStateChange(event) {
        this.setState({
            stickyStatus: event.status,
        });
    }

    _onTriggerNotice(message) {
        this.setState({ message }, this._registerRemovalTimer);
    }

    _updateNoticeVisibility() {
        this._timer = -1;

        this.setState({
            message: '',
        });
    }

    render() {
        const message = this.state.message != null
            ? this.state.message
            : '';

        if (message.length === 0) {
            return null;
        }

        return (
            <div className={'notice-container'}>
                <Sticky
                    enabled={true}
                    top={0}
                    onStateChange={this._onStickyStateChangeHandler}
                >
                    <div className={this._buildClassnames()}>
                        <Notice message={message} isVisible={this.state.message !== ''} />
                    </div>
                </Sticky>
            </div>
        );
    }
}

NoticeContainer.propTypes = {
    message: PropTypes.string,
};

NoticeContainer.defaultProps = {
    message: '',
};
