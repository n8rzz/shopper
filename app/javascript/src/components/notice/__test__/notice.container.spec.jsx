import React from 'react';
import { mount, shallow } from 'enzyme';

import { NoticeContainer } from '../notice.container';

const messageMock = 'this is a notice!';

describe('NoticeContainer', () => {
    describe('when called with an empty #mesage prop', () => {
        test('renders correctly', () => {
            let component = shallow(<NoticeContainer
                message={''}
            />);

            expect(component).toMatchSnapshot();

            component = shallow(<NoticeContainer
                message={null}
            />);

            expect(component).toMatchSnapshot();
        });
    });

    describe('when called with #message', () => {
        let component = null;

        beforeEach(() => {
            component = shallow(<NoticeContainer
                message={messageMock}
            />);
        });

        afterEach(() => {
            component = null;
        });

        test('renders correctly with valid props', () => {
            expect(component).toMatchSnapshot();
        });

        test('#_timer should not equal -1', () => {
            expect(component.instance()._time).not.toBe(-1);
        });
    });

    describe('`stickyHeader`', () => {
        let component = null;

        beforeEach(() => {
            component = mount(<NoticeContainer
                message={messageMock}
            />);
        });

        afterEach(() => {
            component = null;
        });

        describe('when `stickyHeader` is not sticky', () => {
            test('._onStickyStateChange() returns correct classnames', () => {
                const stickyEventMock = { status: 0 };

                component.instance()._onStickyStateChange(stickyEventMock);
                component.update();

                expect(component.find('.js-notice').hasClass('mix-stickyHeader_isSticky')).toBe(false);
            });
        });

        describe('when `stickyHeader` is sticky', () => {
            test('._onStickyStateChange() returns correct classnames', () => {
                const stickyEventMock = { status: 2 };

                component.instance()._onStickyStateChange(stickyEventMock);
                component.update();

                expect(component.find('.js-notice').hasClass('mix-stickyHeader_isSticky')).toBe(true);
            });
        });
    });
});
