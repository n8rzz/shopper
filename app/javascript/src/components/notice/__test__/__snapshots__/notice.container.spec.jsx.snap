// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoticeContainer when called with #message renders correctly with valid props 1`] = `
<div
  className="notice-container"
>
  <Sticky
    activeClass="active"
    bottomBoundary={0}
    enableTransforms={true}
    enabled={true}
    onStateChange={[Function]}
    releasedClass="released"
    shouldFreeze={[Function]}
    top={0}
  >
    <div
      className="js-notice"
    >
      <Notice
        isVisible={true}
        message="this is a notice!"
      />
    </div>
  </Sticky>
</div>
`;

exports[`NoticeContainer when called with an empty #mesage prop renders correctly 1`] = `""`;

exports[`NoticeContainer when called with an empty #mesage prop renders correctly 2`] = `""`;
